src/saltext/vault/states/vault_pki.py

```diff
@@ -0,0 +1,490 @@
+
+    issuer
+        Issuer reference for the role. Can be name, id or literal ``default``.
+
+    ttl
+        Specifies the Time To Live value to be used for the validity period of the requested certificate,
+        provided as a string duration with time suffix. Hour is the largest suffix.
+        The value specified is strictly used for future validity.
+        If not set, uses the system default value or the value of ``max_ttl``, whichever is shorter.
+
+    max_ttl
+        Specifies the maximum Time To Live provided as a string duration with time suffix.
+        Hour is the largest suffix. If not set, defaults to the system maximum lease TTL.
+
+    kwargs
+        Any other parameter accepted by Vault
+        :obj:`write_role <saltext.vault.modules.vault_pki.write_role>` execution module or Vault update role API method.
+    """
+
+    ret = {
+        "name": name,
+        "result": True,
+        "comment": "The role is present as specified",
+        "changes": {},
+    }
+
+    kwargs = {k: v for k, v in kwargs.items() if not k.startswith("_")}
+
+    def _diff_params(current):
+        nonlocal issuer_ref, ttl, max_ttl, kwargs
+        diff_params = (
+            ("issuer_ref", issuer_ref),
+            ("ttl", timestring_map(ttl, cast=int)),
+            ("max_ttl", timestring_map(max_ttl, cast=int)),
+        )
+        changed = {}
+        for param, arg in diff_params:
+            if arg is None:
+                continue
+            if current[param] != arg:
+                changed.update(
+                    {
+                        param: {
+                            "old": current.get(param),
+                            "new": arg,
+                        }
+                    }
+                )
+        for param, arg in kwargs.items():
+            if param in current and current[param] != arg:
+                changed.update(
+                    {
+                        param: {
+                            "old": current.get(param),
+                            "new": arg,
+                        }
+                    }
+                )
+        return changed
+
+    current = __salt__["vault_pki.read_role"](name, mount=mount)
+    changes = {}
+
+    try:
+        if current:
+            changes = _diff_params(current)
+            if not changes:
+                return ret
+
+        if __opts__["test"]:
+            ret["result"] = None
+            ret["comment"] = (
+                f"PKI role `{name}` would have been {'updated' if current else 'created'}"
+            )
+            ret["changes"].update(changes)
+            if not current:
+                ret["changes"]["created"] = name
+            return ret
+
+        __salt__["vault_pki.write_role"](
+            name=name, mount=mount, issuer_ref=issuer_ref, ttl=ttl, max_ttl=max_ttl, **kwargs
+        )
+
+        if not current:
+            ret["changes"]["created"] = name
+
+        ret["changes"].update(changes)
```

Comment:
suggestion: Since the changes are reset on failure of the request, you can do it like this:

```suggestion
        ret["changes"].update(changes)
        if not current:
            ret["changes"]["created"] = name
        if __opts__["test"]:
            ret["result"] = None
            ret["comment"] = (
                f"PKI role `{name}` would have been {'updated' if current else 'created'}"
            )
            return ret

        __salt__["vault_pki.write_role"](
            name=name, mount=mount, issuer_ref=issuer_ref, ttl=ttl, max_ttl=max_ttl, **kwargs
        )
```

Reply:
Looks better...Reworked.